developer/system/gui.rst

```diff
@@ -119,1422 +117,309 @@ Each message starts with the following header:
+                        uint32_t cursor;
+                   };
 
-.. raw:: html
+          - Update cursor pointer for a window. Supported cursor IDs are default cursor (0) and `X Font cursors <https://tronche.com/gui/x/xlib/appendix/b/>`__ (with 0x100 bit set).
 
-   <td>
+        * - MSG_WMCLASS
+          - .. code:: c
 
-Change window state request; fields contains bitmask which flags request to be set and which unset
+                   struct msg_wmclass {
+                       char res_class[64];
+                       char res_name[64];
+                   };
 
-.. raw:: html
+          - Set the WM_CLASS property of a window.
 
-   </td>
+        * - MSG_WINDOW_DUMP
+          - .. code:: c
 
-.. raw:: html
+                   struct msg_window_dump_hdr {
+                       uint32_t type;
+                       uint32_t width;
+                       uint32_t height;
+                       uint32_t bpp;
+                   };
 
-   </tr>
+          - Header for shared memory dump command of type hdr.type. Currently only ``WINDOW_DUMP_TYPE_GRANT_REFS``  (0) is supported.
+        * -     WINDOW_DUMP_TYPE_GRANT_REFS
+          -     .. code:: c
 
-.. raw:: html
+                    struct msg_window_dump_grant_refs {
+                        uint32_t refs[0];
+                    };
 
-   <tr>
 
-.. raw:: html
+          -     Grant references that should be mapped into the compositing buffer.
 
-   <td>
 
-MSG_CURSOR
 
-.. raw:: html
 
-   </td>
+GuiVM -> AppVM messages
+-----------------------
 
-.. raw:: html
+Proper handling of the below messages is NOT security-critical.
 
-   <td>
+Each message starts with the following header
 
-.. raw:: html
+.. code:: c
 
-   <pre>
-   struct msg_cursor {
-        uint32_t cursor;
+   struct msghdr {
+           uint32_t type;
+           uint32_t window;
    };
-   </pre>
 
-.. raw:: html
-
-   </td>
-
-.. raw:: html
-
-   <td>
-
-Update cursor pointer for a window. Supported cursor IDs are default cursor (0) and `X Font cursors <https://tronche.com/gui/x/xlib/appendix/b/>`__ (with 0x100 bit set).
-
-.. raw:: html
-
-   </td>
-
-.. raw:: html
-
-   </tr>
-
-.. raw:: html
-
-   <tr>
-
-.. raw:: html
-
-   <td>
-
-MSG_WMCLASS
-
-.. raw:: html
+The header is followed by message-specific data:
 
-   </td>
+.. list-table::
 
-.. raw:: html
+        * - Message name
+          - Structure after header
+          - Action
 
-   <td>
+        * - MSG_KEYPRESS
+          - .. code:: c
 
-.. raw:: html
+                   struct msg_keypress {
+                     uint32_t type;
+                     uint32_t x;
+                     uint32_t y;
+                     uint32_t state;
+                     uint32_t keycode;
+                   };
 
-   <pre>
-   struct msg_wmclass {
-       char res_class[64];
-       char res_name[64];
-   };
-   </pre>
+          - Tell ``qubes_drv`` driver to generate a keypress
+
+        * - MSG_BUTTON
+          - .. code:: c
```

Comment:
This cell has not been rendered. Maybe due to a missing empty line after `.. code:: c`?